crates/ghost_actor/src/ghost_channel.rs

```diff
@@ -0,0 +1,271 @@
+            self.sender
+                .send(GhostChannelMessage::Response {
+                    request_id: request_id.clone(),
+                    payload,
+                })
+                .expect("should send");
+        }
+    }
+}
+
+pub struct GhostChannel<
+    RequestToOther,
+    RequestToOtherResponse,
+    RequestToSelf,
+    RequestToSelfResponse,
+    Error,
+> {
+    sender: crossbeam_channel::Sender<
+        GhostChannelMessage<RequestToOther, RequestToSelfResponse, Error>,
+    >,
+    receiver: crossbeam_channel::Receiver<
+        GhostChannelMessage<RequestToSelf, RequestToOtherResponse, Error>,
+    >,
+}
+
+impl<RequestToOther, RequestToOtherResponse, RequestToSelf, RequestToSelfResponse, Error>
+    GhostChannel<
+        RequestToOther,
+        RequestToOtherResponse,
+        RequestToSelf,
+        RequestToSelfResponse,
+        Error,
+    >
+{
+    fn new(
+        sender: crossbeam_channel::Sender<
+            GhostChannelMessage<RequestToOther, RequestToSelfResponse, Error>,
+        >,
+        receiver: crossbeam_channel::Receiver<
+            GhostChannelMessage<RequestToSelf, RequestToOtherResponse, Error>,
+        >,
+    ) -> Self {
+        Self { sender, receiver }
+    }
+
+    pub fn as_context_channel<Context>(
+        self,
+    ) -> GhostContextChannel<
+        Context,
+        RequestToOther,
+        RequestToOtherResponse,
+        RequestToSelf,
+        RequestToSelfResponse,
+        Error,
+    > {
+        GhostContextChannel::new(self.sender, self.receiver)
+    }
+}
+
+pub struct GhostContextChannel<
+    Context,
+    RequestToOther,
+    RequestToOtherResponse,
+    RequestToSelf,
+    RequestToSelfResponse,
+    Error,
+> {
+    sender: crossbeam_channel::Sender<
+        GhostChannelMessage<RequestToOther, RequestToSelfResponse, Error>,
+    >,
+    receiver: crossbeam_channel::Receiver<
+        GhostChannelMessage<RequestToSelf, RequestToOtherResponse, Error>,
+    >,
+    await_request_to_other_response: GhostTracker<Context, RequestToOtherResponse, Error>,
+    outbox_requests_to_self:
+        Vec<GhostMessage<RequestToSelf, RequestToOther, RequestToSelfResponse, Error>>,
+}
+
+impl<
+        Context,
+        RequestToOther,
+        RequestToOtherResponse,
+        RequestToSelf,
+        RequestToSelfResponse,
+        Error,
+    >
+    GhostContextChannel<
+        Context,
+        RequestToOther,
+        RequestToOtherResponse,
+        RequestToSelf,
+        RequestToSelfResponse,
+        Error,
+    >
+{
+    fn new(
+        sender: crossbeam_channel::Sender<
+            GhostChannelMessage<RequestToOther, RequestToSelfResponse, Error>,
+        >,
+        receiver: crossbeam_channel::Receiver<
+            GhostChannelMessage<RequestToSelf, RequestToOtherResponse, Error>,
+        >,
+    ) -> Self {
+        Self {
+            sender,
+            receiver,
+            await_request_to_other_response: GhostTracker::new("uuuh?"),
```

Comment:
uuuh?  :-)

I guess in should be a parameter of new, or generated via an auto-incremented lazy_static?